src/encoder/partition.rs

```diff
@@ -0,0 +1,557 @@
+        / ((1 << OD_BITRES) as f64);
+    }
+
+    let partitions = [
+      bo,
+      &BlockOffset { x: bo.x + hbs as usize, y: bo.y },
+      &BlockOffset { x: bo.x, y: bo.y + hbs as usize },
+      &BlockOffset { x: bo.x + hbs as usize, y: bo.y + hbs as usize }
+    ];
+    rd_cost += partitions
+      .iter()
+      .map(|&offset| {
+        encode_partition_bottomup(
+          seq,
+          fi,
+          fs,
+          cw,
+          w_pre_cdef,
+          w_post_cdef,
+          subsize,
+          offset,
+          pmvs //&best_decision.mvs[0]
+        )
+      })
+      .sum::<f64>();
+
+    // Recode the full block if it is more efficient
+    if !must_split && nosplit_rd_cost < rd_cost {
+      rd_cost = nosplit_rd_cost;
+
+      cw.rollback(&cw_checkpoint);
+      w_pre_cdef.rollback(&w_pre_checkpoint);
+      w_post_cdef.rollback(&w_post_checkpoint);
+
+      partition = PartitionType::PARTITION_NONE;
+
+      if bsize >= BlockSize::BLOCK_8X8 {
+        let w: &mut dyn Writer =
+          if cw.bc.cdef_coded { w_post_cdef } else { w_pre_cdef };
+        cw.write_partition(w, bo, partition, bsize);
+      }
+
+      // FIXME: redundant block re-encode
+      let (mode_luma, mode_chroma) =
+        (best_decision.pred_mode_luma, best_decision.pred_mode_chroma);
+      let cfl = best_decision.pred_cfl_params;
+      let ref_frames = best_decision.ref_frames;
+      let mvs = best_decision.mvs;
+      let skip = best_decision.skip;
+      let mut cdef_coded = cw.bc.cdef_coded;
+      let (tx_size, tx_type) = (best_decision.tx_size, best_decision.tx_type);
+
+      debug_assert!(
+        (tx_size, tx_type)
+          == rdo_tx_size_type(
+            seq, fi, fs, cw, bsize, bo, mode_luma, ref_frames, mvs, skip
+          )
+      );
+      cw.bc.set_tx_size(bo, tx_size);
+
+      let mut mv_stack = Vec::new();
+      let is_compound = ref_frames[1] != NONE_FRAME;
+      let mode_context = cw.find_mvrefs(
+        bo,
+        ref_frames,
+        &mut mv_stack,
+        bsize,
+        false,
+        fi,
+        is_compound
+      );
+
+      cdef_coded = encode_block_a(
+        seq,
+        fs,
+        cw,
+        if cdef_coded { w_post_cdef } else { w_pre_cdef },
+        bsize,
+        bo,
+        skip
+      );
+      encode_block_b(
+        seq,
+        fi,
+        fs,
+        cw,
+        if cdef_coded { w_post_cdef } else { w_pre_cdef },
+        mode_luma,
+        mode_chroma,
+        ref_frames,
+        mvs,
+        bsize,
+        bo,
+        skip,
+        seq.bit_depth,
+        cfl,
+        tx_size,
+        tx_type,
+        mode_context,
+        &mv_stack,
+        false
+      );
+    }
+  }
+
+  subsize = bsize.subsize(partition);
+
+  if bsize >= BlockSize::BLOCK_8X8
+    && (bsize == BlockSize::BLOCK_8X8
+      || partition != PartitionType::PARTITION_SPLIT)
+  {
+    cw.bc.update_partition_context(bo, subsize, bsize);
+  }
+
+  rd_cost
+}
+
+pub fn encode_partition_topdown(
```

Comment:
These two partition rdo functions are core of rdo based mode decision. So, I think that it is better in the previous file name rdo.rs.